sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java

```diff
@@ -95,6 +96,20 @@ public static boolean isMutable(DataType dt) {
       pdt instanceof PhysicalCalendarIntervalType;
   }
 
+  /**
+   * True if comparisons, equality and hashing can be done purely on binary representation of
+   * Unsafe row. i.e. binary(e1) = binary(e2) <=> e1 = e2.
+   * e.g. this is not true for non-binary collations (any case/accent insensitive collation
+   * can lead to rows being semantically equal even though their binary representations differ).
+   */
+  public static boolean isBinaryStable(DataType dt) {
+    if (dt instanceof StringType st) {
```

Comment:
shall we check nested field/array elements recursively?

Reply:
oh, I see. Yes, we need that as well.

Reply:
done. I also moved method to `UnsafeRowUtils`. It feels a bit more natural to do this kind of checks in scala.